### ISSUE TITLE
Clarify the context of InvalidGlobal

### ISSUE THREAD

**Issue by Rarst** [vimeo/psalm#4120]
If a global variable declaration is encountered in a typical PHP template file the error is emitted:

> ERROR: InvalidGlobal - ... - Cannot use global scope here (see https://psalm.dev/046)

And that URL says:

> Emitted when there's a reference to the global keyword where it's not expected
>
> ```php
> <?php
>
> global $e;
> ```

I am puzzled by this and it might be missing sufficient context?

1. "Cannot" why? It is valid PHP syntactically and functionally.
2. "Not expected" by what? PHP engine is fine with it.

I am not advocating _for_ globals here, but I don't see what about this is an _error_ exactly.

**Comment by azjezz** [comment]
Because using global variables causes very tight coupling of code, and mutates the global state of your application, which means if any function/method/class/file in your application decided to change the value of the said global variable ( using `$GLOBALS['foo'] = $bar;` for example ), the output of your template will changes.

a package installed using composer can affect the behavior of your application as well due to the usage of global variables ( as it can mutate the value ).

therefore, they are banned ( +psalm cannot keep track of the type of this variable since it can be changed anywhere ).

most modern PL (especially strong typed language ), ban/forbid the usage of global variables, and Psalm is doing the right thing by banning it as well.

**Comment by Rarst** [comment]
My point is, as worded, that doesn't express "globals in general are wrong" it says "something with _this_ global is wrong".

I don't disagree with your explanation, but that isn't what sniff name/message/documentation communicates at the moment.

**Comment by muglug** [comment]
Yeah, I suppose it could be removed because (as you say) it can be valid here:

`a.php`:

```php
<?php

$a = 5;

function foo() {
  require('b.php');
}
```

`b.php`
```php
<?php

global $a;

echo $a;
```
